### ISSUE TITLE
Disable PowerShell autosaving context

### ISSUE THREAD

**Issue by doherty100** [Azure-Samples/azuresandbox#56]
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [ X ] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

### Minimal steps to reproduce
> Provision #AzureSandbox

### Any log messages given by the failure
> `WARNING: The provided service principal secret or certifcate password will be included in the 'keystore.cache' file found in the user profile ( /home/rdoherty/.Azure ). Please ensure that this directory has appropriate protections.`

### Expected/desired behavior
> Remediate this warning by disabling autosaving of context in PowerShell scripts.

### OS and Version?
> N/A

### Versions
> #AzureSandbox v2.10.1

### Mention any other details that might be useful

> This needs to be implemented in all PowerShell scripts that are used in the bootstrapping process.

**Comment by doherty100** [comment]
## Analysis: is this caching disableable, and how would it affect the repo?

This caching is **by design**, and several repo workflows depend on it. Summary below.

### Why the warning appears
The `keystore.cache` warning is emitted specifically when authenticating with a **service principal secret** via `Connect-AzAccount -ServicePrincipal -Credential`. Az PowerShell persists the SP secret (not just a token) into `~/.Azure/keystore.cache` so the context survives between sessions. It's the context-autosave feature, not a defect.

### Yes, there are ways to disable it

**Azure PowerShell** (per Microsoft docs):
- `Disable-AzContextAutosave` — stops persisting credentials to `~/.Azure`; the login is forgotten when the session ends. It does **not** clear already-stored tokens (use `Clear-AzContext` / `Logout-AzAccount` for that).
- `Connect-AzAccount -Scope Process` — the context is valid for the current process only and is never written to disk, **regardless of the autosave setting**. This is the cleanest per-invocation fix.

**Azure CLI**: there is **no** "never persist" toggle. The CLI always writes an MSAL token cache to `~/.azure` by design; you can only clear it after the fact (`az logout`, `az account clear`). (`az config param-persist off` relates to *command parameter* persistence, not credentials.)

### How disabling would impact this repo
The "connected, session-to-session" experience is real and baked into several workflows:

1. **`Invoke-UnitTests.ps1`** calls `Get-AzContext` and hard-fails if no persisted context exists. Each test run is a fresh `pwsh -File …` invocation, so a global `Disable-AzContextAutosave` would force re-authentication on every run.
2. **`bootstrap.ps1`** reads `Get-AzContext` / `Get-AzAccessToken` to derive defaults — depends on a live/persisted login.
3. **Azure CLI** (`az login`) persistence is required by `terraform apply` and `enable-public-access.sh` across separate command invocations.

So a blanket disable would break the fire-and-forget setup model and force interactive re-auth mid-run.

### Where does SP-secret auth actually happen? (full inventory, excluding Terraform's own provider auth)

**Active runtime SP-secret login:**
- **`modules/vnet-app/scripts/Set-AzureFilesConfiguration.ps1`** — runs on the domain controller during `vnet_app` provisioning to configure identity-based (Kerberos) access for Azure Files. It first logs in with `Connect-AzAccount -Identity` (managed identity) to read the SP secret from Key Vault, then re-authenticates with `Connect-AzAccount -ServicePrincipal -Credential` because the AD/CIFS SPN registration must run as the SP. **This is the one active runtime workflow that needs the SP secret, and the site most likely to emit the `keystore.cache` warning** on a Linux/pwsh host.

**Active one-shot helper:**
- **`scripts/Create-AzSqlDbUser.ps1`** — authenticates as the SP (SQL admin) to mint a DB access token and create a contained DB user. Already holds the secret via a parameter; single run.

**Retired (not part of the current sandbox apply):**
- `extras/modules/retired/vm-devops-win/scripts/Register-DscNode.ps1`
- `extras/modules/retired/vm-devops-win/scripts/Set-AutomationAccountConfiguration.ps1`

**Not SP-secret logins (for clarity):**
- The **bootstrap scripts** (`scripts/bootstrap.ps1`, `extras/configurations/rg-devops-iac/scripts/bootstrap.ps1`) authenticate **interactively** (`Connect-AzAccount -UseDeviceAuthentication`) and only reference the SP via a `Get-AzADServicePrincipal -AppId` *lookup* — no secret is cached, no warning.
- All other **VM-side and integration-test scripts** (`Invoke-MssqlConfiguration.ps1`, `Set-MssqlConfiguration.ps1`, `Invoke-AzureFilesConfiguration.ps1`, `Test-*.ps1`) use `Connect-AzAccount -Identity` (managed identity) — no secret written, no warning.

### Recommendation — scope the fix to the SP-secret call sites (don't disable autosave globally)

Disabling autosave everywhere breaks the interactive/test UX above. Instead, remediate per script. The right lever differs depending on whether the script performs **one** login or **multiple** logins in a single process:

**Single-login, one-shot scripts → add `-Scope Process` to the `Connect-AzAccount` call.**
- `scripts/Create-AzSqlDbUser.ps1` (and the retired scripts) do exactly one SP login, acquire a token, and exit. `-Scope Process` keeps the context in memory only, so nothing is written to `~/.Azure` and the warning never fires. Zero functional impact — the script never relies on cross-session persistence.

**Multi-login scripts → use `Disable-AzContextAutosave -Scope Process` once at the top instead.**
- `modules/vnet-app/scripts/Set-AzureFilesConfiguration.ps1` performs **two** logins in one process: `Connect-AzAccount -Identity` (managed identity, default scope) and then `Connect-AzAccount -ServicePrincipal` (SP secret, default scope). Adding `-Scope Process` to *only* the SP call would leave the earlier managed-identity login autosaving its context to `~/.Azure` — mixed scopes, and only a partial fix. A single `Disable-AzContextAutosave -Scope Process` at the top of the script covers **both** logins, guarantees nothing (MI context or SP secret) is written to disk for the life of the run, and — because this script also runs one-shot on the DC via `azurerm_virtual_machine_run_command` — has no functional impact. (Equivalently, you could add `-Scope Process` to *both* `Connect-AzAccount` calls; the single `Disable-AzContextAutosave` line is tidier.)
- Note: `-Scope Process` / `Disable-AzContextAutosave` only governs the **on-disk** `keystore.cache` write, which is what the warning targets. In this script the secret is also pulled from Key Vault into a plaintext in-memory variable (`Get-AzKeyVaultSecret -AsPlainText`); that in-memory copy is unaffected and is a separate consideration.

Either way, the persisted-context convenience is preserved for `bootstrap.ps1`, `Invoke-UnitTests.ps1`, and the `az login` / Terraform flows, because the fix is applied only to the one-shot SP-secret scripts.

**Comment by doherty100** [comment]
## Proposed changes (final plan)

Decision: rather than patch the service-principal (SP) login in `Set-AzureFilesConfiguration.ps1`, **remove it at the source** by letting `jumpwin1`'s managed identity perform the storage operations directly. This eliminates both the on-disk `keystore.cache` write (the original issue) and the in-memory plaintext secret, and lets us drop the SP secret from Key Vault entirely. `Set-AzureFilesConfiguration.ps1` runs one-shot on `jumpwin1` (the domain-joined Windows jumpbox), so there's no cross-session persistence to preserve.

### 1. Grant `jumpwin1`'s managed identity the storage control-plane role it currently lacks
The script's SP login exists only because `jumpwin1`'s MI can't run these control-plane operations:
- `New-AzStorageAccountKey` → `Microsoft.Storage/storageAccounts/regeneratekey/action`
- `Get-AzStorageAccountKey -ListKerbKey` → `.../listkeys/action`
- `Set-AzStorageAccount -EnableActiveDirectoryDomainServicesForFile` → `Microsoft.Storage/storageAccounts/write`

Its MI has only **Storage Blob Data Reader** (read-only data-plane) on the storage account. Add a control-plane role — **Storage Account Contributor** (or a custom role limited to `listkeys`/`regeneratekey`/`write`), scoped to the storage account — to the `vm_win_roles` map in `modules/vnet-app/locals.tf`, and order the role assignment ahead of the `Set-AzureFilesConfiguration` run command.

### 2. Simplify `Set-AzureFilesConfiguration.ps1` to a single managed-identity session
- Remove the "read SP secret from Key Vault → `ConvertTo-SecureString` → `Connect-AzAccount -ServicePrincipal`" block.
- Perform all storage operations under the existing `Connect-AzAccount -Identity` session.
- Remove now-unused script parameters (`AppId`, and `TenantId`/`SubscriptionId` if no longer needed) and update the run-command wiring in `modules/vnet-app/locals.tf` accordingly.

### 3. Optional tidiness — don't persist the managed-identity context either
Even after the SP login is gone, `Connect-AzAccount -Identity` still writes context/tokens (not a secret) to `~/.Azure`. Add a single `Disable-AzContextAutosave -Scope Process` at the top of the script (or `-Scope Process` on the `-Identity` login) so nothing from this one-shot run is persisted. This is cleanup, not the issue-56 warning fix (that's already resolved by removing the SP login).

### 4. Remove the now-orphaned SP secret from Key Vault
Once step 2 lands, `azurerm_key_vault_secret.spn_password` (in `modules/vnet-shared/main.tf`, named after the SP appId) has **zero remaining consumers** (verified: `Set-AzureFilesConfiguration.ps1` was its only reader; no Terraform `data` source reads it). So:
- Delete the `azurerm_key_vault_secret "spn_password"` resource.
- Remove the `secret_spn_password` line from the `key_vault_operations_complete` barrier signal.
- Remove the now-unused `arm_client_secret_version` variable. **Keep `arm_client_secret`** — it's still required for provider auth and by `Create-AzSqlDbUser.ps1`.

### 5. Vestigial cleanup (unrelated, opportunistic)
`modules/vm-mssql-win/scripts/Invoke-MssqlConfiguration.ps1` declares an `AppId` parameter it never uses, and `modules/vm-mssql-win/locals.tf` passes it in. Neither reads the SP secret. Drop both.

### Explicitly out of scope
`scripts/Create-AzSqlDbUser.ps1` still performs an SP-secret login, but it sources the secret from `TF_VAR_arm_client_secret` (a `protected_parameter` in root `main.tf`), **not** from Key Vault — so removing the KV secret does not affect it. If desired, its own `keystore.cache` warning can be silenced separately with `-Scope Process` on its `Connect-AzAccount` call. Tracked as a possible follow-up, not part of this change.

### Docs & validation
- Update the `vnet-app` module README and the root README RBAC/variables/resources tables to reflect the new role assignment and the removed `spn_password` secret / `arm_client_secret_version` variable.
- Run `./scripts/Invoke-CIChecks.sh terraform powershell` (and `markdown` for the README edits) clean before opening/pushing the PR to `vnext`.

### Net effect
- Original issue resolved: no SP secret is ever written to `keystore.cache`, because there is no SP login in `Set-AzureFilesConfiguration.ps1` anymore.
- No plaintext SP secret in memory on `jumpwin1`.
- The SP secret is no longer stored in Key Vault at all.
- Trade-off (accepted): `jumpwin1`'s MI gains a standing storage control-plane role. This was weighed against the previous design and preferred, since the prior SP approach still allowed a compromised `jumpwin1` to read the SP secret from Key Vault (standing `Key Vault Secrets User`) and assume the SP's broader Contributor/Owner rights — a wider latent capability than a storage-scoped role.

**Comment by doherty100** [comment]
## Security trade-off documented (privilege expansion vs. not persisting an Owner-scoped SPN secret)

For the record, the fix in #646 makes a deliberate trade-off. Both sides:

**Cost — jumpwin1's managed identity gains a standing privilege.**
- Before: jumpwin1's MI held only `Key Vault Secrets User` (KV), `Storage Blob Data Reader` and (test-only) `Monitoring Reader`.
- After: it additionally holds **`Storage Account Contributor`**, scoped to the single sandbox storage account. This is a real, *standing* control-plane grant on the VM identity (key regeneration + enabling AD DS auth on that storage account), where previously the VM identity had no storage control-plane rights.

**Benefit — we stop persisting a highly privileged SPN secret.**
- The old design copied the deployment **service principal's secret into Key Vault** (`spn_password`) and had jumpwin1 read it and log in as the SPN. That SPN is the Terraform deployment identity and carries **subscription-scoped Owner-level RBAC** — far broader than anything the sandbox VM should be able to wield.
- Persisting that secret created two exposures the warning in this issue pointed at: it was written to disk (`~/.Azure/keystore.cache`) and materialized as plaintext in memory on jumpwin1, and it sat at rest in Key Vault for the life of the sandbox.
- Crucially, the old approach did **not** actually contain the VM's blast radius: jumpwin1 already had standing `Key Vault Secrets User`, so a compromise of jumpwin1 could read the Owner-scoped SPN secret from Key Vault and assume Owner over the subscription. That is a strictly *wider* latent capability than the new standing role.

**Net assessment.**
- We trade a **narrow, storage-account-scoped standing role** on the VM identity for the **removal of a subscription-Owner secret** from disk, memory, and Key Vault.
- Reducing the worst-case blast radius from "assume subscription Owner" to "regenerate keys / enable AD DS auth on one storage account" is a clear net security improvement, even though the VM identity's *nominal* standing footprint grows by one role. Least-standing-privilege on the VM identity is slightly worse; least-*worst-case*-privilege (and secret sprawl) is materially better.

**Out of scope (noted for completeness).** `Create-AzSqlDbUser.ps1` still logs in as the SPN, but it sources the secret from `TF_VAR_arm_client_secret` (a `protected_parameter`), **not** from Key Vault — so it neither persists the secret at rest nor is affected by this change. Silencing its transient `keystore.cache` write with `-Scope Process` remains a possible follow-up.

**Comment by doherty100** [comment]
Resolved by #646 (merged to `vnext`). Azure Files identity-based access is now configured via the jumpwin1 managed identity instead of a service-principal login, so the `spn_password` secret is no longer persisted in Key Vault or handled in memory/on disk, and the `keystore.cache` warning is eliminated (the remaining runtime login uses `Disable-AzContextAutosave -Scope Process`). Validated end-to-end with a fresh sandbox deployment: 282 resources applied and the full test suite passed 109/109.
